src/ucp/core/ucp_worker.c

```diff
@@ -1168,6 +1168,60 @@ void ucp_worker_iface_cleanup(ucp_worker_iface_t *wiface)
     ucp_worker_uct_iface_close(wiface);
 }
 
+static ucs_status_t ucp_worker_add_resource_cms(ucp_worker_h worker)
+{
+    ucp_context_h   context  = worker->context;
+    ucs_status_t    status   = UCS_OK;
+    uct_cm_h        *cms_tmp;
+    ucp_rsc_index_t i;
+
+    if (!ucp_worker_sockaddr_is_cm_proto(worker)) {
+        return UCS_ERR_UNSUPPORTED;
+    }
+
+    cms_tmp = ucs_alloca(context->num_cmpts);
+    ucs_assert_always(cms_tmp != NULL);
+
+    UCS_ASYNC_BLOCK(&worker->async);
+    worker->num_cms = 0;
+    for (i = 0; i < context->num_cmpts; ++i) {
+        status = uct_cm_open(context->tl_cmpts[i].cmpt, worker->uct,
+                             &cms_tmp[worker->num_cms]);
+        if (status == UCS_OK) {
+            ++worker->num_cms;
+        } else {
+            ucs_assert_always(status == UCS_ERR_UNSUPPORTED);
```

Comment:
we should check UCT component flags for CM support, rather than relying on return status
(this way we can also count how many CMs we have before allocating, would not need to use alloca)